deltacat/compute/compactor/steps/materialize.py

```diff
@@ -164,49 +172,66 @@ def _materialize(
-        pa_table = sc.append_dedupe_task_idx_col(
-            pa_table,
-            dedupe_task_indices,
-        )
-        pa_sort_keys = [(sc._DEDUPE_TASK_IDX_COLUMN_NAME, "ascending")]
-        pa_table = pa_table.take(
-            pc.sort_indices(pa_table, sort_keys=pa_sort_keys),
-        )
-        pa_table = pa_table.drop(
-            [sc._DEDUPE_TASK_IDX_COLUMN_NAME]
-        )
+        if remainder_tables:
+            assert remainder_record_count < max_records_per_output_file, \
+                f"Total records in previous remainder should not exceed {max_records_per_output_file}."
+
+            # Skip materializing tables until we have reached 'max_records_per_output_file'
+            if remainder_record_count + len(pa_table) < max_records_per_output_file:
+                remainder_tables.append(pa_table)
+                remainder_record_count += len(pa_table)
+                continue
+
+            # 'max_records_per_output_file' will be exceeded with the
+            #  current file's records + previous remainder records.
+            records_to_fit = max_records_per_output_file - remainder_record_count
+            fitted_table = pa_table.slice(length=records_to_fit)
+            remainder_tables.append(fitted_table)
 
-        # Write manifests up to max_records_per_output_file
-        # TODO(raghumdani): Write exactly the same number of records into each file to
-        # produce a read-optimized view of the tables.
-        if compacted_tables and \
-                total_record_count + record_count > max_records_per_output_file:
-            materialized_results.append(_materialize(compacted_tables, total_record_count))
+            # Chop off records that are already written to disk
+            pa_table = pa_table.slice(offset=records_to_fit)
+
+        record_count = len(pa_table)
+        assert record_count > 0, f"Expected non-empty compacted PyArrow table"
+        record_multiplier, leftover = \
+            record_count // max_records_per_output_file, record_count % max_records_per_output_file
+
+        # Pack as many records divisible by max_records_per_output_file into one stage delta call
+        if record_multiplier > 0:
+            fitted_remainder_table = pa_table.slice(length=record_multiplier * max_records_per_output_file)
+            remainder_tables.append(fitted_remainder_table)
+
+        if remainder_tables:
+            # Materialize tables up to this point which are divisible by max_records_per_output_file
+            materialized_results.append(_materialize(remainder_tables))
             # Free up written tables in memory
-            compacted_tables.clear()
-            total_record_count = 0
+            remainder_tables.clear()
+            remainder_record_count = 0
 
-        total_record_count += record_count
-        compacted_tables.append(pa_table)
+        # Add any leftovers (less than max_records_per_output_file) to the remainder list after packing
```

Comment:
Instead of adding comments what the code next to it is doing (which is fairly easy to understand), what you may add is a doc to this method and explain `why` part i.e., why does it have to add left overs to remainder table.

Reply:
How about a separate issue altogether for adding docstrings to each of the compaction steps?

Reply:
Makes sense.